client/lib/browser-storage/index.ts

```diff
@@ -0,0 +1,137 @@
+/**
+ * Internal dependencies
+ */
+import {
+	getStoredItem as bypassGet,
+	setStoredItem as bypassSet,
+	clearStorage as bypassClear,
+} from './bypass';
+
+let shouldBypass = false;
+
+const DB_NAME = 'calypso';
+const DB_VERSION = 2; // Match versioning of the previous localforage-based implementation.
+const STORE_NAME = 'calypso_store';
+
+const supportsIDB = typeof window !== 'undefined' && window.indexedDB;
+
+let cachedDbPromise: Promise< IDBDatabase > | null = null;
+
+function prepareIDB() {
+	const request = indexedDB.open( DB_NAME, DB_VERSION );
+	return new Promise< IDBDatabase >( ( resolve, reject ) => {
+		if ( request ) {
+			request.onerror = () => reject( request.error );
+			request.onsuccess = () => resolve( request.result );
+			request.onupgradeneeded = () => request.result.createObjectStore( STORE_NAME );
+		}
+	} );
+}
+
+function getDB() {
+	cachedDbPromise = cachedDbPromise || prepareIDB();
+	return cachedDbPromise;
```

Comment:
If you don't mind using Lodash, I like to use this elegant pattern:
```js
const getDB = _.once( () => new Promise( ( resolve, reject ) => {
  ...
} ) );

getDb();
getDb(); // second invocation returns the result of the first
```

Reply:
I'd rather avoid lodash for such a simple job, particularly since `once` is one of those methods that ends up pulling in an unreasonable portion of lodash because it relies on `before`, which unnecessarily does type casting (`toInteger` -> `toFinite` -> `toNumber` -> `isSymbol` -> `isObjectLike` and `baseGetTag`, etc.).

How about:
```
const getDB = () => ( cachedDbPromise = cachedDbPromise || prepareIDB() );
```

Reply:
Actually, there's `memoizeLast` in the codebase already. I'll use that.